sql/core/src/main/scala/org/apache/spark/sql/expressions/ReduceAggregator.scala

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.expressions
+
+import org.apache.spark.annotation.Experimental
+import org.apache.spark.sql.Encoder
+import org.apache.spark.sql.catalyst.encoders.ExpressionEncoder
+
+/**
+ * :: Experimental ::
+ * An aggregator that uses a single associative and commutative reduce function. This reduce
+ * function can be used to go through all input values and reduces them to a single value.
+ * If there is no input, a null value is returned.
+ *
+ * @since 2.1.0
+ */
+@Experimental
+abstract class ReduceAggregator[T] extends Aggregator[T, (Boolean, T), T] {
+
+  // Question 1: Should func and encoder be parameters rather than abstract methods?
+  //  rxin: abstract method has better java compatibility and forces naming the concrete impl,
+  //  whereas parameter has better type inference (infer encoders via context bounds).
+  // Question 2: Should finish throw an exception or return null if there is no input?
+  //  rxin: null might be more "SQL" like, whereas exception is more Scala like.
```

Comment:
i prefer null over an exception

ideally this would return an option. the operation really should be:
```def reduceGroups(f: (V, V) => V): Dataset[(K, Option[V])]```
and the class:
```abstract class ReduceAggregator[T] extends Aggregator[T, (Boolean, T), Option[T]]```